backend/src/database/state-migrations/migrations.ts

```diff
@@ -52,55 +54,59 @@ export const migrations: {
-        targetStateVersion
-    );
-    const stateMigrationFunctions = migrationFunctions.state;
-    const originalStates = await entityManager.findOne(ExerciseWrapperEntity, {
+    const exercise = await entityManager.findOne(ExerciseWrapperEntity, {
         where: { id: exerciseId },
-        select: { initialStateString: true, currentStateString: true },
     });
-    if (originalStates === null) {
+    if (exercise === null) {
         throw new RestoreError(
             'Cannot find exercise to convert in database',
             exerciseId
         );
     }
-    let initialState: object | null = null,
-        currentState: object | null = null;
-    if (stateMigrationFunctions.length > 0) {
-        initialState = JSON.parse(originalStates.initialStateString);
-        currentState = JSON.parse(originalStates.currentStateString);
-        stateMigrationFunctions.forEach((stateMigrationFunction) => {
-            initialState = stateMigrationFunction(initialState!);
-            currentState = stateMigrationFunction(currentState!);
-        });
-    }
-    const actionsMigrationFunctions = migrationFunctions.actions;
-    let actions: (object | null)[] | null = null;
-    if (actionsMigrationFunctions.length > 0) {
-        const originalActions = await entityManager.find(ActionWrapperEntity, {
+    const targetVersion =
+        typeof targetStateVersion === 'number'
+            ? targetStateVersion
+            : Math.max(...StrictObject.keys(migrations));
+    if (targetStateVersion <= exercise.stateVersion) return;
+    let initialState = JSON.parse(exercise.initialStateString);
+    let currentState = JSON.parse(exercise.currentStateString);
+    const actions = (
+        await entityManager.find(ActionWrapperEntity, {
             where: { exercise: { id: exerciseId } },
             select: { actionString: true },
             order: { index: 'ASC' },
-        });
-        actions = originalActions.map((originalAction) =>
-            JSON.parse(originalAction.actionString)
-        );
-        initialState ??= JSON.parse(originalStates.initialStateString);
-        actionsMigrationFunctions.forEach((actionsMigrationFunction) => {
-            actionsMigrationFunction(initialState!, actions!);
-        });
+        })
+    ).map((action) => JSON.parse(action.actionString));
+    for (
+        let currentStateVersion = exercise.stateVersion + 1;
+        currentStateVersion <= targetVersion;
+        currentStateVersion++
+    ) {
+        const stateMigration = migrations[currentStateVersion]?.state;
+        if (typeof stateMigration === 'function') {
+            initialState = stateMigration(initialState);
+            currentState = stateMigration(currentState);
+        }
+        if (!actions) continue;
+        const actionMigration = migrations[currentStateVersion]?.actions;
+        if (typeof actionMigration === 'function') {
+            actionMigration(initialState, actions);
+        }
     }
```

Comment:
This should be an extra function as it is reused in the other migration function.